### DISCUSSION TITLE
Persistence for Open Source Edition

### DISCUSSION THREAD

**Discussion by erezblm** [emqx/emqx#10846] [Q&A] [answered]
Hi all,

I was a bit confused from the various materials I read regarding persistence in EMQX, would be glad if you could help me.
For clarifying my use case, I'm deploying EMQX 5.0 open source to Azure AKS through the EMQX k8s operator.

Questions:

1. I was reading that persistence in disk is supported only in enterprise edition, what is included in this limitation? QOS 1/2 disk persistence? session disk persistence? retained messages disk persistence?
2. in the enterprise edition, where is the data stored? in the mria db? is it replicated through instances?
3. If persistence is not applied in open source so what is the usage of the mria database?
4. Regarding the new upcoming EMQX 5.1 RocksDB support, will it be open to the open source edition?
5. Is the new RocksDB db replacing the mria model? or is it unrelated?
6. What is the usage of the cloud storage (in my case Azure Storage)

Appreciate your support
Erez.

**Comment by thalesmg** [comment, marked as answer]
Hi @erezblm .

1. The enterprise edition (EE) supports persistence through its [data bridges](https://docs.emqx.com/en/enterprise/v5.0/data-integration/data-bridges.html).  The community edition (CE) only supports webhook and MQTT bridges.
   1. Retained messages do have an option to use Mnesia on-disk persistence in the CE edition.
2. Data in the case of bridges is stored in the remote databases.
3. [mria](https://github.com/emqx/mria/) is used to improve the scalability of EMQX clusters by introducing two types of nodes: cores and replicants.  It still uses Mnesia as its backend, so any `disc_copies` or `disc_only_copies` tables will be written to disk.  The session/message tables in EMQX are not in this category, though: they are currently `ram_copies`.
4. Yes.
5. RocksDB will be added as a storage option which mria may use, similar to `disc_copies` and `ram_copies`.  So it'll just augment mria.
6. Once you start using EMQX with RocksDB support in a cloud environment, you'll probably want to use a persistence volume, so I guess that's where Azure Storage would come in.
